Handle crc status json output that now returns output to stdOut, fixe…

odockal opened the pull request:
…s JBIDE-27727

Signed-off-by: Ondrej Dockal <odockal@redhat.com>

# Pull Request Checklist
## General
* Is this a blocking issue or new feature? If yes, QE needs to +1 this PR

## Code
* Are method-/class-/variable-names meaningful?
* Are methods concise, not too long?
* Are catch blocks catching precise Exceptions only (no catch all)?

## Testing
* Are there unit-tests?
* Are there integration tests (or at least a jira to tackle these)?
* Is the non-happy path working, too?
* Are other parts that use the same component still working fine?

## Function
* Does it work?

jeffmaury commented on plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/core/adapter/CRC100Poller.java:
```diff
@@ -153,8 +153,28 @@ protected IStatus runOnePingJson(IServer server, Map<String, String> env)
+			    	if (actualObj.has("error")) {
+					    JsonNode errorField = actualObj.get("error");
+					    String error = errorField.asText();
+				    	if (error.contains("Machine does not exist")) {
```
What is the error message is different ? We may endup with the same NPE as before

odockal replied:
Error message did not change. But the `crc status -o json` did. Before it sent this kind of message to the stdErr. Now it is on stdOut. There are other options as of what can error could possibly contain, so I have to check that it not yet started crc like "error". 

There are multiple options that could fill in error field: https://github.com/code-ready/crc/blob/f51090d9f0086b822daf41503a16978d826c0b67/cmd/crc/cmd/status.go#L49.

jeffmaury replied:
Why don't you simply trust the success=false flag so that we will be error free when crc decided to change their error message

odockal replied:
Because I think that simply trusting success=false does not always indicate that machine is in "not yet started, but otherwise alright" state. I believe that there are some situations where you should not let adapter into "starting" state, but this needs more details from crc side. If it would be addressed in sprint to come I am ok with that.

jkopriva commented:
@jeffmaury you are right: It didi not work for me, I have very similar NPE(row numbers are different):
`!ENTRY org.eclipse.core.jobs 4 2 2021-03-22 18:05:28.668
!MESSAGE An internal error occurred during: "Starting CodeReady Containers 1.0+".
!STACK 0
java.lang.NullPointerException
	at org.jboss.tools.openshift.internal.crc.server.core.adapter.CRC100Poller.shouldUseJsonFlags(CRC100Poller.java:57)
	at org.jboss.tools.openshift.internal.crc.server.core.adapter.CRC100Poller.onePing(CRC100Poller.java:104)
	at org.jboss.tools.openshift.internal.cdk.server.core.adapter.AbstractCDKPoller.onePingSafe(AbstractCDKPoller.java:100)
	at org.jboss.tools.openshift.internal.cdk.server.core.adapter.AbstractCDKPoller.getCurrentStateSynchronous(AbstractCDKPoller.java:88)
	at org.jboss.tools.openshift.internal.crc.server.core.adapter.controllers.CRC100LaunchController.launch(CRC100LaunchController.java:219)
	at org.jboss.ide.eclipse.as.wtp.core.server.launch.ControllableServerLaunchConfiguration.launch(ControllableServerLaunchConfiguration.java:52)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:803)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:716)
	at org.eclipse.debug.internal.core.LaunchConfiguration.launch(LaunchConfiguration.java:711)
	at org.eclipse.wst.server.core.internal.Server.startImpl2(Server.java:3568)
	at org.eclipse.wst.server.core.internal.Server.startImpl(Server.java:3504)
	at org.eclipse.wst.server.core.internal.Server$StartJob.run(Server.java:379)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)`

odockal commented:
> `!ENTRY org.eclipse.core.jobs 4 2 2021-03-22 18:05:28.668 !MESSAGE An internal error occurred during: "Starting CodeReady Containers 1.0+". !STACK 0 java.lang.NullPointerException at org.jboss.tools.openshift.internal.crc.server.core.adapter.CRC100Poller.shouldUseJsonFlags(CRC100Poller.java:57) at org.jboss.tools.openshift.internal.crc.server.core.adapter.CRC100Poller.onePing(CRC100Poller.java:104) at org.jboss.tools.openshift.internal.cdk.server.core.adapter.AbstractCDKPoller.onePingSafe(AbstractCDKPoller.java:100) at org.jboss.tools.openshift.internal.cdk.server.core.adapter.AbstractCDKPoller.getCurrentStateSynchronous(AbstractCDKPoller.java:88) at 

The line of NPE suggests it came from crcVersion handling (`crcVers.split`)...
```
	protected boolean shouldUseJsonFlags() {
		String[] versionSegments = crcVers.split("\\.");
		int[] versionSegmentsInt = new int[versionSegments.length];
```
What version do you use? Did you remove .crc before testing this? Did you call setup crc and overrided .crc?

jeffmaury reviewed (approved):
LGTM

odockal commented:
@jeffmaury @jkopriva While trying to debug and understand code more, I also suddenly experience NPE Josef got... It is coming from deriving minishift version function. I will try to debug it.

jkopriva commented:
@odockal it works for me now!

odockal commented:
I suspect timeout while getting the version from binary...

sbouchet reviewed (approved):
LGTM.

odockal commented:
I tried to reproduce the issue by setting some cycles and catching possible exceptions, it all failed. So I did not reproduce. I think it was very rare situation.
